test/integration/lib/lockfile.spec.ts

```diff
@@ -0,0 +1,200 @@
+import { expect } from 'chai';
+import { promises as fs, mkdirSync } from 'fs';
+import { testfs, TestFs } from 'mocha-pod';
+import * as os from 'os';
+import * as path from 'path';
+import { stub } from 'sinon';
+import * as lockfile from '~/lib/lockfile';
+import * as fsUtils from '~/lib/fs-utils';
+
+const NOBODY_UID = 65534;
+
+describe('lib/lockfile', () => {
+	const lockdir = '/tmp/lockdir';
+
+	let testFs: TestFs.Enabled;
+
+	beforeEach(async () => {
+		testFs = await testfs(
+			{
+				[lockdir]: {
+					'other.lock': testfs.file({ uid: NOBODY_UID }),
+				},
+			},
+			{ cleanup: [path.join(lockdir, '**.lock')] },
+		).enable();
+	});
+
+	afterEach(async () => {
+		await testFs.restore();
+	});
+
+	it('should create a lockfile as the current user by default', async () => {
+		const lock = path.join(lockdir, 'updates.lock');
+
+		// Take the lock passing a uid
+		await expect(lockfile.lock(lock)).to.not.be.rejected;
+
+		// The file should exist
+		await expect(fs.access(lock)).to.not.be.rejected;
+
+		// Verify lockfile UID
+		expect((await fs.stat(lock)).uid).to.equal(os.userInfo().uid);
+	});
+
+	it('should create a lockfile as the `nobody` user at target path', async () => {
+		const lock = path.join(lockdir, 'updates.lock');
+
+		// Take the lock passing a uid
+		await expect(lockfile.lock(lock, NOBODY_UID)).to.not.be.rejected;
+
+		// The file should exist
+		await expect(fs.access(lock)).to.not.be.rejected;
+
+		// Verify lockfile UID
+		expect((await fs.stat(lock)).uid).to.equal(NOBODY_UID);
+	});
+
+	it('should not be able to take the lock if it already exists', async () => {
+		const lock = path.join(lockdir, 'updates.lock');
+
+		// Take the lock passing a uid
+		await expect(lockfile.lock(lock, NOBODY_UID)).to.not.be.rejected;
+
+		// The file should exist
+		await expect(fs.access(lock)).to.not.be.rejected;
+
+		// Trying to take the lock again should fail
+		await expect(lockfile.lock(lock, NOBODY_UID)).to.be.rejected;
+	});
+
+	it('should create a lockfile with the provided `uid` if specified', async () => {
+		const lock = path.join(lockdir, 'updates.lock');
+
+		// Take the lock passing a uid
+		await expect(lockfile.lock(lock, 2)).to.not.be.rejected;
+
+		// The file should exist
+		await expect(fs.access(lock)).to.not.be.rejected;
+
+		// Verify lockfile UID
+		expect((await fs.stat(lock)).uid).to.equal(2);
+	});
+
+	it('should not create a lockfile if `lock` throws', async () => {
+		// Stub the call to exec.
+		// WARNING: This is relying on internal knowledge of the function
+		// which is generally not a good testing practice, but I'm not sure
+		// how to do it otherwise
```

Comment:
Yeah, it also bugged me that I could not figure out how to stub `child_process.exec` and had to stub `fs-utils.exec` instead. Since `fs-utils.exec` uses a promisified version of the former, then stubs were never being used. Anyway is not such a big deal now.